File: engine/src/main/java/ru/vertuos/engine/map/object/InclineObject.java
package ru.vertuos.engine.map.object;

import ru.vertuos.engine.world.object.DynamicGameObject;

public class InclineObject extends CollisionableGameObject {

    public InclineObject(float positionX, float positionY, float width, float height) {
        super(positionX, positionY, width, height);
    }

    @Override
    public void onCollide(DynamicGameObject dynamic) {
        float deltaX = positionX - dynamic.getPositionX();
        float deltaY = deltaX * height / width;
        dynamic.setPositionY(positionY - deltaY);
        dynamic.setOnGround(true);
    }
}

File: engine/src/main/java/ru/vertuos/engine/map/InclineObject.kt
package ru.vertuos.engine.map

import ru.vertuos.engine.math.Vector2f
import ru.vertuos.engine.util.setPositionY
import ru.vertuos.engine.world.obj.DynamicGameObject

class InclineObject(
    position: Vector2f, size: Vector2f
) : CollisionableGameObject(position, size) {

    override fun onCollide(obj: DynamicGameObject) {
        val deltaX: Float = position.x - obj.position.x
        val deltaY: Float = deltaX * size.y / size.x
        obj.setPositionY(position.y - deltaY)
        obj.isOnGround = true
    }
}
